core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java

```diff
@@ -434,82 +439,102 @@ public void setMaxRequestLength(long maxRequestLength) {
         final HttpHeaders additionalResponseHeaders = this.additionalResponseHeaders;
         if (additionalResponseHeaders == null) {
-            return HttpHeaders.EMPTY_HEADERS;
+            return HttpHeaders.of();
         }
-        return additionalResponseHeaders.asImmutable();
+        return additionalResponseHeaders;
     }
 
     @Override
-    public void setAdditionalResponseHeader(AsciiString name, String value) {
+    public void setAdditionalResponseHeader(CharSequence name, Object value) {
         requireNonNull(name, "name");
         requireNonNull(value, "value");
-        createAdditionalHeadersIfAbsent().set(name, value);
+        additionalResponseHeaders = createAdditionalHeadersIfAbsent().toBuilder()
+                                                                     .setObject(name, value).build();
     }
 
     @Override
-    public void setAdditionalResponseHeaders(Headers<? extends AsciiString, ? extends String, ?> headers) {
+    public void setAdditionalResponseHeaders(Iterable<? extends Entry<? extends CharSequence, ?>> headers) {
         requireNonNull(headers, "headers");
-        createAdditionalHeadersIfAbsent().set(headers);
+        additionalResponseHeaders = createAdditionalHeadersIfAbsent().toBuilder().setObject(headers).build();
     }
 
     @Override
-    public void addAdditionalResponseHeader(AsciiString name, String value) {
+    public void addAdditionalResponseHeader(CharSequence name, Object value) {
         requireNonNull(name, "name");
         requireNonNull(value, "value");
-        createAdditionalHeadersIfAbsent().add(name, value);
+        additionalResponseHeaders = createAdditionalHeadersIfAbsent().toBuilder()
+                                                                     .addObject(name, value).build();
     }
 
     @Override
-    public void addAdditionalResponseHeaders(Headers<? extends AsciiString, ? extends String, ?> headers) {
+    public void addAdditionalResponseHeaders(Iterable<? extends Entry<? extends CharSequence, ?>> headers) {
         requireNonNull(headers, "headers");
-        createAdditionalHeadersIfAbsent().add(headers);
+        additionalResponseHeaders = createAdditionalHeadersIfAbsent().toBuilder().addObject(headers).build();
     }
 
     @Override
-    public boolean removeAdditionalResponseHeader(AsciiString name) {
+    public boolean removeAdditionalResponseHeader(CharSequence name) {
         requireNonNull(name, "name");
-        return createAdditionalHeadersIfAbsent().remove(name);
+        final HttpHeaders additionalResponseHeaders = this.additionalResponseHeaders;
+        if (additionalResponseHeaders == null || additionalResponseHeaders.isEmpty()) {
+            return false;
+        }
+
+        final HttpHeadersBuilder builder = createAdditionalHeadersIfAbsent().toBuilder();
+        final boolean removed = builder.remove(name);
+        this.additionalResponseHeaders = builder.build();
```

Comment:
ditto

Reply:
Updated `*HeadersBuilder` so that `build()` returns the original headers if no changes were made.